sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -384,22 +502,55 @@ public void initialize(PipelineOptions options) throws Exception {}
      * @param writerResults the results of writes (FileResult).
      */
     @Override
-    public void finalize(Iterable<FileResult> writerResults, PipelineOptions options)
+    public void finalize(Iterable<FileResult> writerResults,
+                         PipelineOptions options)
         throws Exception {
       // Collect names of temporary files and rename them.
-      List<String> files = new ArrayList<>();
-      for (FileResult result : writerResults) {
-        LOG.debug("Temporary bundle output file {} will be copied.", result.getFilename());
-        files.add(result.getFilename());
-      }
-      copyToOutputFiles(files, options);
+      Map<String, String> outputFilenames = buildOutputFilenames(writerResults);
+      copyToOutputFiles(outputFilenames, options);
 
+      // Optionally remove temporary files.
       // We remove the entire temporary directory, rather than specifically removing the files
       // from writerResults, because writerResults includes only successfully completed bundles,
       // and we'd like to clean up the failed ones too.
       // Note that due to GCS eventual consistency, matching files in the temp directory is also
       // currently non-perfect and may fail to delete some files.
-      removeTemporaryFiles(files, options);
+      //
+      // When windows or triggers are specified, files are generated incrementally so deleting
+      // the entire directory in finalize is incorrect.
+      removeTemporaryFiles(outputFilenames.keySet(), !windowedWrites, options);
+    }
+
+    protected final Map<String, String> buildOutputFilenames(Iterable<FileResult> writerResults) {
+      Map<String, String> outputFilenames = new HashMap<>();
+      List<String> files = new ArrayList<>();
+      for (FileResult result : writerResults) {
+        if (result.getDestinationFilename() != null) {
+          outputFilenames.put(result.getFilename(), result.getDestinationFilename());
+        } else {
+          files.add(result.getFilename());
+        }
+      }
+
+      // If the user does not specify numShards() (not supported with windowing). Then the
```

Comment:
It took me a minute to figure out this comment. So this whole code path is dead in streaming, hence it is OK to hard-code the fact that no window or pane info is passed?